src/App.tsx

```diff
@@ -1,26 +1,226 @@
+        handleError(Errors.DELETING, setErrorNotification);
+      })
+      .finally(() => {
+        setLoadingTodoIds(currentIds =>
+          currentIds.filter(currentId => currentId !== todoId),
+        );
+        setFocusOnChange(current => !current);
+      });
+  };
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!inputChange) {
+      handleError(Errors.EMPTY, setErrorNotification);
+
+      return;
+    }
+
+    const newTempTodo = {
+      title: inputChange.trim(),
+      userId: USER_ID,
+      completed: false,
+      id: 0,
+    };
+
+    setTempTodo(newTempTodo);
+
+    const newTodo = {
+      title: inputChange.trim(),
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setLoadingTodoIds(currentIds => [...currentIds, 0]);
+
+    addTodo(newTodo)
+      .then(res => {
+        setTodos(currentTodos => [...currentTodos, res]);
+        setInputChange('');
+      })
+      .catch(() => {
+        handleError(Errors.ADDING, setErrorNotification);
+      })
+      .finally(() => {
+        setLoadingTodoIds(currentIds =>
+          currentIds.filter(currentId => currentId !== 0),
+        );
+        setTempTodo(null);
+        setFocusOnChange(current => !current);
+      });
+  };
+
+  const toggleTodo = (todoId: number, updates: Partial<Todo>) => {
+    setLoadingTodoIds(currentIds => [...currentIds, todoId]);
+
+    updateTodo(todoId, updates)
+      .then(res => {
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            res.id === currentTodo.id ? res : currentTodo,
+          ),
+        );
+      })
+      .catch(() => {
+        handleError(Errors.UPDATING, setErrorNotification);
+      })
+      .finally(() => {
+        setLoadingTodoIds(currentIds =>
+          currentIds.filter(currentId => currentId !== todoId),
+        );
+        setFocusOnChange(current => !current);
+      });
+  };
+
+  const handleRenameTodo = (
+    todoId: number,
+    updateChanges: string,
+    todoTitle: string,
+  ) => {
+    if (!updateChanges.length) {
+      handleDelete(todoId);
+
+      return;
+    }
+
+    if (updateChanges === todoTitle) {
+      setEditingTodo(null);
+
+      return;
+    }
+
+    setLoadingTodoIds(currentIds => [...currentIds, todoId]);
+
+    setTodos(currentTodos =>
+      currentTodos.map(todo =>
+        todo.id === todoId ? { ...todo, title: updateChanges } : todo,
+      ),
+    );
+
+    const updates = {
+      title: updateChanges.trim(),
+    };
+
+    updateTodo(todoId, updates)
+      .then(updatedTodo => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+        setEditingTodo(null);
+      })
+      .catch(() => {
+        handleError(Errors.UPDATING, setErrorNotification);
+        setTodos(currentTodos =>
+          currentTodos === todos ? currentTodos : todos,
+        );
```

Comment:
it's not necessary, but in your case I would change only title for needed todo, because you have access to that (almost the same as on line 150)